### PR TITLE
sql: extend filter propagation to groupNode.

### PR DESCRIPTION
**Pull request by knz**
Filters on non-aggregated renders of groupNode can be propagated.

---
This change is [Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/13194)

### CODE REVIEW

**Comment by RaduBerinde** [comment]
Nice! :lgtm:

---

Reviewed 3 of 3 files at r1.
Review status: all files reviewed at latest revision, 1 unresolved discussion, some commit checks pending.

---

*[pkg/sql/testdata/aggregate, line 1010 at r1](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KbVhEesf-8NIdyxOpb_:-KbVhEesf-8NIdyxOpba:b-a4p5cb) ([raw file](https://github.com/cockroachdb/cockroach/blob/826d8cb7baa94232ab5afa887b265c9c2bfb2713/pkg/sql/testdata/aggregate#L1010)):*
> ```
> 2          table        kv@primary
> 2          spans        ALL
> 2          filter       1 > 10
> ```

not due to this change but strange formatting, should probably be `@2 > 10`

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-:-KbVhRnIvA8Sv6_snudD:be8rotp)*

**Comment by knz** [comment]
Ugh, actually the test shows that the implementation is currently plainly wrong. The indexedvars in the aggFuncHolders of a groupNode are messed up. Will work on this again after I deal with #12525 further.

**Comment by knz** [comment]
Case in point:
```
explain(exprs,metadata,symvars) selecT * from (SELECT max(k+v)+max(k),v FROM t.kv GROUP BY v) where v > 10;
+-------+--------+-------------+------------------------+----------------------------+-----------+
| Level |  Type  |    Field    |      Description       |          Columns           | Ordering  |
+-------+--------+-------------+------------------------+----------------------------+-----------+
|     0 | group  |             |                        | ("max(k + v) + max(k)", v) |           |
|     0 |        | aggregate 0 | max(@1 + @2)           |                            |           |
|     0 |        | aggregate 1 | max(@1)                |                            |           |
|     0 |        | aggregate 2 | @2                     |                            |           |
|     0 |        | render 0    | max(@1 + @2) + max(@1) |                            |           |
|     0 |        | render 1    | @2                     |                            |           |
|     1 | render |             |                        | ("k + v", k, v, v)         | +k,unique |
|     1 |        | render 0    | @1 + @2                |                            |           |
|     1 |        | render 1    | @1                     |                            |           |
|     1 |        | render 2    | @2                     |                            |           |
|     1 |        | render 3    | @2                     |                            |           |
|     2 | scan   |             |                        | (k, v)                     | +k,unique |
|     2 |        | table       | kv@primary             |                            |           |
|     2 |        | spans       | /11-                   |                            |           |
|     2 |        | filter      | @1 > 10                |                            |           |
+-------+--------+-------------+------------------------+----------------------------+-----------+
```

**Comment by RaduBerinde** [comment]
Wow, that's clearly messed up. A bit troubling that we don't have tests that catch this. I wonder if this commit screwed things up: https://github.com/cockroachdb/cockroach/commit/b4edfaeaf484c8824668e40ce7311af85e30c73a

**Comment by knz** [comment]
There's no functional error though because groupNode never evaluates the arguments of its aggFuncHolders. But it's so bizarrely correct that the entire thing is ripe for a redesign though.

RaduBerinde <notifications@github.com> schreef op 28 januari 2017 19:27:55 CET:
>Wow, that's clearly messed up. A bit troubling that we don't have tests
>that catch this. I wonder if this commit screwed things up:
>https://github.com/cockroachdb/cockroach/commit/b4edfaeaf484c8824668e40ce7311af85e30c73a
>
>-- 
>You are receiving this because you authored the thread.
>Reply to this email directly or view it on GitHub:
>https://github.com/cockroachdb/cockroach/pull/13194#issuecomment-275865345

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by RaduBerinde** [comment]
With the recent work for #12525, this will need some refactoring but it should work fine now.

**Comment by knz** [comment]
Yes funny you say that, I just started looking at it 20 minutes ago :)

**Comment by knz** [comment]
Done, PTAL

**Comment by knz** [comment]
Nice to see how the distsql plan is simplified automatically :)

**Comment by RaduBerinde** [comment]
Ooops, missed your previous update, sorry about that.

Very nice! :lgtm_strong: 

---

Review status: 0 of 5 files reviewed at latest revision, 8 unresolved discussions, all commit checks successful.

---

*[pkg/sql/filter_opt.go, line 391 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLer_o0tgTrDGTxX79:-KkLer_o0tgTrDGTxX7A:b-z0bh2n) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L391)):*
> ```Go
> // addGroupFilter attempts to add the extraFilter to the groupNode.
> // The filter is only propagated to the sub-plan if it is expressed
> // using simple column references to the source.
> ```

I would rephrase to "The part of the filter that depends only on GROUP BY expressions is propagated to the source". ("column references" is a bit iffy because if you have `GROUP BY a+b`, the `a+b` is a "column" (render) in the source but it's not really a column as an expression).

---

*[pkg/sql/filter_opt.go, line 395 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLe1Q7z-D_LBTTVGvs:-KkLe1Q7z-D_LBTTVGvt:b5imvae) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L395)):*
> ```Go
> 	ctx context.Context, g *groupNode, info *dataSourceInfo, extraFilter parser.TypedExpr,
> ) (planNode, parser.TypedExpr, error) {
> 	// outerFilter is the filter expressed using values rendered by this groupNode.
> ```

[nit] I would avoid the use of "rendered" for groupNode results, since it's not doing arbitrary renderings any more. Maybe just "values generated by" or "results of the groupNode"

---

*[pkg/sql/filter_opt.go, line 396 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLefw3PqXz737aGwaN:-KkLefw4nth5BEvk6Ahs:b-pbr1yh) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L396)):*
> ```Go
> ) (planNode, parser.TypedExpr, error) {
> 	// outerFilter is the filter expressed using values rendered by this groupNode.
> 	var outerFilter parser.TypedExpr = parser.DBoolTrue
> ```

[nit] It would be a bit more clear if we initialized this to `extraFilter` (or just renamed `extraFilter` to `outerFilter`), so we don't have to think about the special "if extraFilter is true, we leave this as true" case. Feel free to ignore if this doesn't make sense.

---

*[pkg/sql/filter_opt.go, line 401 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLeFMRhZQAROlKyLca:-KkLeFMRhZQAROlKyLcb:b-aoermk) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L401)):*
> ```Go
> 
> 	if !isFilterTrue(extraFilter) {
> 		// The filter that's being added refers to the rendered
> ```

[nit] no "rendered"

---

*[pkg/sql/filter_opt.go, line 403 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLeK_dqAttpasWAdpP:-KkLeK_dqAttpasWAdpQ:bav4dpq) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L403)):*
> ```Go
> 		// The filter that's being added refers to the rendered
> 		// expressions, not the render's source node.
> 		convFunc := func(v parser.VariableExpr) (bool, parser.Expr) {
> ```

Maybe add `// We can push (part of) the filter down through IDENT "aggregations"`

---

*[pkg/sql/group.go, line 577 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLgHYVIfX9_P3efcGe:-KkLgHYWxB2XHBRni9Hg:b-44x9ir) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/group.go#L577)):*
> ```Go
> 		}
> 		f := v.groupNode.newAggregateFuncHolder(
> 			v.preRender.render[groupIdx], groupIdx, true, parser.NewIdentAggregate,
> ```

true /* ident */

---

*[pkg/sql/group.go, line 610 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLgJiMA0h9cKN46ry9:-KkLgJiMA0h9cKN46ryA:b-76enx7) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/group.go#L610)):*
> ```Go
> 			argRenderIdx := v.preRender.addOrReuseRender(col, argExpr, true /* reuse */)
> 
> 			f := v.groupNode.newAggregateFuncHolder(t, argRenderIdx, false, agg)
> ```

false /* not ident */

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-:-KkLg_dI2EiD6OQC7ysM:b-gcki3k)*

**Comment by RaduBerinde** [comment]
*[pkg/sql/testdata/logic_test/explain_types, line 86 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLi0V9dbq1141RXE7r:-KkLi0V9dbq1141RXE7s:b-6kh0rj) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/testdata/logic_test/explain_types#L86)):*
> ```
> 
> query ITTTTT
> EXPLAIN (TYPES,NOEXPAND) SELECT 2*COUNT(k) as z, v FROM t WHERE v>123 GROUP BY v HAVING v<2
> ```

I would add a test where we actually have to split the `HAVING` filter (e.g. `HAVING v <500 AND z >3`).

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194)*

**Comment by knz** [comment]
thanks :)

---

Review status: 0 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[pkg/sql/filter_opt.go, line 391 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLer_o0tgTrDGTxX79:-KkLiELu4SgY7L-Jvpnu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L391)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I would rephrase to "The part of the filter that depends only on GROUP BY expressions is propagated to the source". ("column references" is a bit iffy because if you have `GROUP BY a+b`, the `a+b` is a "column" (render) in the source but it's not really a column as an expression).
</blockquote></details>

Done.

---

*[pkg/sql/filter_opt.go, line 395 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLe1Q7z-D_LBTTVGvs:-KkLiJcURC2h6mCjDn9F:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L395)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] I would avoid the use of "rendered" for groupNode results, since it's not doing arbitrary renderings any more. Maybe just "values generated by" or "results of the groupNode"
</blockquote></details>

Done.

---

*[pkg/sql/filter_opt.go, line 396 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLefw3PqXz737aGwaN:-KkLin8MtxJOP9ZKvbhE:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L396)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] It would be a bit more clear if we initialized this to `extraFilter` (or just renamed `extraFilter` to `outerFilter`), so we don't have to think about the special "if extraFilter is true, we leave this as true" case. Feel free to ignore if this doesn't make sense.
</blockquote></details>

Done.

---

*[pkg/sql/filter_opt.go, line 401 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLeFMRhZQAROlKyLca:-KkLinaF9SRGckL-91Gk:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L401)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

[nit] no "rendered"
</blockquote></details>

Done.

---

*[pkg/sql/filter_opt.go, line 403 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLeK_dqAttpasWAdpP:-KkLjZra9QLX4OMdRjBa:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/filter_opt.go#L403)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

Maybe add `// We can push (part of) the filter down through IDENT "aggregations"`
</blockquote></details>

Done.

---

*[pkg/sql/group.go, line 577 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLgHYVIfX9_P3efcGe:-KkLjhmPPfypSQBgobJu:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/group.go#L577)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

true /* ident */
</blockquote></details>

Done.

---

*[pkg/sql/group.go, line 610 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLgJiMA0h9cKN46ry9:-KkLjiIcVhzfcq7taeHN:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/group.go#L610)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

false /* not ident */
</blockquote></details>

Done.

---

*[pkg/sql/testdata/logic_test/explain_types, line 86 at r2](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-KkLi0V9dbq1141RXE7r:-KkLk56SWi96okZVXkTe:b-896fix) ([raw file](https://github.com/cockroachdb/cockroach/blob/149b523e445e5c40d32f24a28e2f34b74f8f0fd3/pkg/sql/testdata/logic_test/explain_types#L86)):*
<details><summary><i>Previously, RaduBerinde wrote…</i></summary><blockquote>

I would add a test where we actually have to split the `HAVING` filter (e.g. `HAVING v <500 AND z >3`).
</blockquote></details>

Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/cockroach/13194#-:-KkLk6MMO-81qAh2tDqt:b-1bjpmn)*
